pkg/scheduler/internal/queue/equivalence/eqivalence.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file contains structures that implement equivalence class types.
+// Pods under same controllerRef will be put in the same Class.podSet
+// and will be considered as a unit in the process of scheduling.
+// Equivalence class improves scheduler velocity and responsiveness by
+// continuous scheduling when one is determined schedulable and avoiding checking
+// the schedulability of all of these pods when one is determined unschedulable.
+// This is particularly useful in batch processing.
+
+package equivalence
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubernetes/pkg/scheduler/util"
+)
+
+// ClassMap is the type of a map which has the key by a pod's UID of controllerRef and
+// the value which is a pointer to the Class.
+type ClassMap map[types.UID]*Class
+
+var once sync.Once
+var classMap *ClassMap
+
+// NewClassMap creates an empty classMap which is a map key by a pod's UID of
+// controllerRef and the value is a pointer to the Class.
+func NewClassMap() *ClassMap {
+	once.Do(func() {
+		classMap = new(ClassMap)
+		*classMap = make(map[types.UID]*Class)
+	})
+	return classMap
+}
+
+// Class holds pods under same controllerRef and is the unit in scheduling process.
+// of scheduling.
+type Class struct {
+	// pod's UID of controllerRef
+	Hash types.UID
+	// Pods which under same controllerRef waiting to be scheduled.
+	PodSet *sync.Map
+}
+
+// NewClass returns equivalence class if exited or create new equivalence class.
+func NewClass(pod *v1.Pod) *Class {
+	equivHash := GetEquivHash(pod)
+	classMap = NewClassMap()
+
+	if _, ok := (*classMap)[equivHash]; !ok {
```

Comment:
Why don't just do:

```suggestion
	if _, ok := classMap[equivHash]; !ok {
		classMap[equivHash] = &Class{
			Hash:   equivHash,
			PodSet: new(sync.Map),
		}
	}
	
	return classMap[equivHash]
```
Map is Golang is always referred by pointer AFAIK.

Reply:
`classMap` is the pointer to `ClassMap`. Type `ClassMap` can support indexing but type `*ClassMap` does not support indexing. So I use `(*classMap)` to get the value which is type `ClassMap`.